include/openPMD/binding/python/Pickle.hpp

```diff
@@ -0,0 +1,91 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * openPMD-api is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with openPMD-api.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "openPMD/backend/Attributable.hpp"
+#include "openPMD/IO/Access.hpp"
+#include "openPMD/Series.hpp"
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <exception>
+#include <string>
+#include <tuple>
+#include <vector>
+
+
+namespace openPMD
+{
+    /** Helper to Pickle Attributable Classes
+     *
+     * @tparam T_Args the types in pybind11::class_ - the first type will be pickled
+     * @tparam T_SeriesAccessor During unpickle, this accesses the object inside
+     *                          a newly constructed series
+     * @param cl the pybind11 class that gets the pickle methods defined
+     * @param seriesAccessor accessor from series to object during unpickling
+     */
+    template< typename... T_Args, typename T_SeriesAccessor >
+    inline void
+    add_pickle(
+       pybind11::class_< T_Args... > & cl,
+       T_SeriesAccessor && seriesAccessor
+    )
+    {
+        namespace py = pybind11;
+
+        // helper: get first class in py::class_ - that's the type we pickle
+        using PickledClass = typename std::tuple_element<
+            0,
+            std::tuple< T_Args... >
+        >::type;
+
+        cl.def(py::pickle(
+                // __getstate__
+                []( const PickledClass &a ) {
+                    // Return a tuple that fully encodes the state of the object
+                    Attributable::MyPath const myPath = a.myPath();
+                    return py::make_tuple( myPath.filePath(), myPath.group );
+                },
+
+                // __setstate__
+                [&seriesAccessor]( py::tuple t ) {
+                    // our tuple has exactly two elements: filePath & group
+                    if (t.size() != 2)
+                        throw std::runtime_error("Invalid state!");
+
+                    std::string const filename = t[0].cast< std::string >();
+                    std::vector< std::string > const group =
+                        t[1].cast< std::vector< std::string > >();
+
+                    // Create a new openPMD Series and keep it alive.
+                    // This is a big hack for now, but it works for our use
+                    // case, which is spinning up remote serial read series
+                    // for DASK.
+                    static auto series = openPMD::Series(
```

Comment:
Not sure how much of a hack this `static` member is to fix up lifetimes, but it seems to work pretty well :)

Reply:
What happens when a user wants to unpickle two different Series within the same application? The `static auto series` will not be constructed a second time and then be stuck with the old file path.
Do you do this for avoiding the Series from going out of scope?
Would something like `return seriesAccessor( std::move( series ), group)` help? This would ensure that the `seriesAccessor` takes no reference.

Reply:
Ah, I think I see the problem.
```cpp
    add_pickle(
        cl,
        [](openPMD::Series & series, std::vector< std::string > const & group ) {
            uint64_t const n_it = std::stoull(group.at(1));
            return series.iterations[n_it].meshes[group.at(3)][group.at(4)];
        }
    )
```
This lets the `Series` go out of scope and returns only the record component..

I think that this is a fundamental design issue with our Python bindings: Python has garbage collection, but we don't use it. Realistically, calling `del series` should *not* be sufficient in Python to destroy the Series, since the `series` should stay active as long as there is an active dependent member like a contained `RecordComponent`.
So, moving away from a C++-centered memory model for our Python API should probably do the trick.

Question into the blue: As a quick fix, is it somehow possible in PyBind to quickly tell the garbage collection that in the above code the `series` and the returned `MeshRecordComponent` depend on each other?

Reply:
> fundamental design issue with our Python bindings: Python has garbage collection, but we don't use it. 

Offline discussed: What we mean is we don't use the possibility to tell the garbage collector dependencies between our object hierarchy yet (in both directions) yet.

> Is it somehow possible in PyBind to quickly tell the garbage collection that in the above code the series and the returned MeshRecordComponent depend on each other? This should at least implement the above solution locally.

Gotta investigate that, would be a nice and clean solution.